### ISSUE TITLE
Incorrect typings in createSlice

### ISSUE THREAD

**Issue by EliseyMartynovSynder** [reduxjs/redux-toolkit#2169]
Hello! I've a problem with `createSlice` `reducers` function typing.
[Codesandbox](https://codesandbox.io/s/fragrant-frog-eqs6oo?file=/src/testSlice.ts)

**The problem**:

I have initial state that has some type - union - some fields or `null`. I also set this type for action `PayloadAction` and returning my payload to let RTK set this state (since I'm not mutating).
BUT... I'm getting an error on `CaseReducer` and it says that I'm trying to assign my type to type `void`. Why? I'm returning typed payload, not void...
![image](https://user-images.githubusercontent.com/101872641/160464318-98b55f55-7da8-4bde-a949-23dd28b06103.png)

**Comment by markerikson** [comment]
@phryneas can correct me here, but I _think_ it's because of the `const initialState: SomeType = null`.  I seem to remember him saying that TS will sometimes over-interpret things like `const whatever = null`.

If you hover over the `createSlice` call, it seems to be treating the state type as `null`, not `SomeType`.

If I change it to:

`const initialState = null as SomeType`

the error goes away.

**Comment by phryneas** [comment]
Yup. This is essentially over-eager flow analysis. TS thinks that at the moment `configureStore` is called, `initialState` is `null`, so it just takes that type into account and not all the possible other types.

[In the future](https://github.com/microsoft/TypeScript/issues/47920), you will be able to write `const initialState = null satisfies SomeType` (and I'm looking forward to the day when I can change that in the docs), but for now you will have to make do with the slightly unfortunate `const initialState = null as SomeType`

**Comment by EliseyMartynovSynder** [comment]
@markerikson, @phryneas wow, thank you guys! That was not really obvious :) And I missed that in the docs somehow..
